### ISSUE TITLE
[NEW RULE] Avoid Heap Allocation after Initialization

### ISSUE THREAD

**Issue by anandmindfire** [mindfiredigital/eslint-plugin-hub#3]
### 1️⃣ Is your rule proposal related to a problem? Please describe

In performance-sensitive or long-running applications, dynamic memory allocation during runtime can cause performance hits, garbage collection pauses, and memory fragmentation. It becomes especially problematic when allocations happen repeatedly inside loops or functions that are called frequently.

For example: *I’m often frustrated when profiling shows unnecessary allocations inside tight loops that could have been avoided with simple buffer reuse.*

### 2️⃣ Describe the solution you'd like

This rule should detect:

* Repeated object/array creation inside loops or frequently-called functions.
* Opportunities to reuse preallocated structures instead of creating new ones on every call.
* Discourage implicit or hidden allocations (like `[].concat(...)` or `new Array(...)` in hot paths).

✅ **Allowed**:

```js
const reusableBuffer = [];

function processIncomingData(newData) {
  reusableBuffer.length = 0;
  reusableBuffer.push(...newData);
  // Use reusableBuffer...
}
```

❌ **Disallowed**:

```js
function handleLargeBatch(batch) {
  for (let i = 0; i < batch.length; i++) {
    const tempArray = [batch[i]];
    // ... processing ...
  }
}
```

### 3️⃣ Describe alternatives you've considered

* Relying on performance profiling tools after-the-fact, but they require deep inspection and don’t prevent the issue up front.
* Manually reviewing hot-paths for allocations—but this becomes difficult in large or collaborative codebases.

### 4️⃣ Example configuration

```json
{
  "rules": {
    "hub/avoid-heap-after-init": ["warn", { "disallowNewInLoops": true }]
  }
}
```

### 5️⃣ Additional context

This rule is especially helpful for:

* Applications where performance predictability matters (e.g., real-time systems, games, data pipelines).
* Avoiding spikes in memory usage or garbage collection in Node.js or frontend JS.
* Reinforcing good practices like buffer reuse and data pooling.

Could be extended to detect `new Object()`, `new Array()`, array literals (`[]`), object literals (`{}`), or even implicit string concatenation inside hot paths.

**Comment by biswalom22** [comment]
@anandmindfire @lakinmindfire This rule is tested and verified with all possible test cases
